src/pudl/metadata/classes.py

```diff
@@ -1748,3 +1762,66 @@ def to_rst(self, top_dir: DirectoryPath, csv_subdir: DirectoryPath, rst_path: st
                 rendered = encoder.to_rst(
                     top_dir=top_dir, csv_subdir=csv_subdir, is_header=header)
                 f.write(rendered)
+
+
+class DatasetteMetadata(Base):
+    """
+    A collection of Data Sources and Resources for metadata export.
+
+    Used to create metadata YAML file to accompany Datasette.
+    """
+
+    data_sources: Dict[str, DataSource]
+    resource_package: Package = Package.from_resource_ids()
+    label_columns: Dict[str, str]
+
+    @classmethod
+    def from_data_source_ids(
+        cls,
+        data_source_ids: Iterable[str] = [
+            'pudl', 'ferc1', 'eia860', 'eia860m', 'eia923'],
+        extra_etl_groups: Iterable[str] = [
+            'entity_eia', 'glue', 'static_eia', 'static'],
+        label_columns: Dict[str, str] = {
+            'plants_entity_eia': 'plant_name_eia',
+            'plants_ferc1': 'plant_name_ferc1',
+            'plants_pudl': 'plant_name_pudl',
+            'utilities_entity_eia': 'utility_name_eia',
+            'utilities_ferc1': 'utility_name_ferc1',
+            'utilities_pudl': 'utility_name_pudl'
+        }
+    ) -> "DatasetteMetadata":
+        """
+        Construct a dictionary of DataSources from data source names.
+
+        Create dictionary of first and last year or year-month for each source.
+
+        Args:
+            data_source_ids: ids of data sources currently included in Datasette
+            label_columns: map from resource to column name for link label
+        """
+        data_sources = {
+            ds_id: DataSource.from_id(ds_id) for ds_id in data_source_ids}
+        resource_ids = []
+        for name in data_source_ids + extra_etl_groups:
+            resource_ids += DataSource(name=name,
+                                       license_raw=LICENSES["us-govt"],
+                                       license_pudl=LICENSES["cc-by-4.0"]).get_resource_ids()
+        resource_package = Package.from_resource_ids(tuple(resource_ids))
+        return cls(
+            data_sources=data_sources,
+            resource_package=resource_package,
+            label_columns=label_columns)
+
+    def to_yaml(self, path: str) -> None:
+        """Output database, table, and column metadata to YAML file."""
+        template = JINJA_ENVIRONMENT.get_template("metadata.yml.jinja")
+        rendered = template.render(
+            license=LICENSES["cc-by-4.0"],
+            data_sources=self.data_sources,
+            package=self.resource_package,
+            label_columns=self.label_columns)
+        if path:
+            Path(path).write_text(rendered)
+        else:
+            sys.stdout.write(rendered)
```

Comment:
The template got a little messier with looping through the `data_sources` as a list instead of a dictionary but I agree that it's probably cleaner to have `data_sources` be a list that's passed into the template.